dan_layer/engine/src/runtime/impl.rs

```diff
@@ -877,12 +932,15 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> RuntimeInte
                         .check_resource_access_rules(ResourceAuthAction::Withdraw, &resource_lock)?;
 
                     let vault_mut = state.get_vault_mut(&vault_lock)?;
-                    let resource_container = match arg {
-                        VaultWithdrawArg::Fungible { amount } => vault_mut.withdraw(amount)?,
-                        VaultWithdrawArg::NonFungible { ids } => vault_mut.withdraw_non_fungibles(&ids)?,
-                        VaultWithdrawArg::Confidential { proof } => vault_mut.withdraw_confidential(*proof)?,
+                    let (resource_container, amount) = match arg {
+                        VaultWithdrawArg::Fungible { amount } => (vault_mut.withdraw(amount)?, Some(amount)),
+                        VaultWithdrawArg::NonFungible { ids } => (vault_mut.withdraw_non_fungibles(&ids)?, None),
+                        VaultWithdrawArg::Confidential { proof } => (vault_mut.withdraw_confidential(*proof)?, None),
```

Comment:
Each withdraw can have an amount. For non-fungible, it's the number of tokens. For Confidential, it can be the number of revealed tokens.